[NFC][LoopVectorize] Make replaceVPBBWithIRVPBB more efficient

david-arm opened the pull request:
In replaceVPBBWithIRVPBB we spend time erasing and appending predecessors and successors from a list, when all we really have to do is replace the old with the new. Not only is this more efficient, but it also preserves the ordering of successors and predecessors. This is something which may become important for vectorising early exit loops (see PR #88385), since a VPIRInstruction is the wrapper for a live-out phi with extra operands that map to the incoming block according to the block's predecessor.

david-arm committed:
[NFC][LoopVectorize] Make replaceVPBBWithIRVPBB more efficient

In replaceVPBBWithIRVPBB we spend time erasing and appending
predecessors and successors from a list, when all we really have
to do is replace the old with the new. Not only is this more
efficient, but it also preserves the ordering of successors and
predecessors. This is something which may become important for
vectorising early exit loops (see PR #88385), since a
VPIRInstruction is the wrapper for a live-out phi with extra
operands that map to the incoming block according to the block's
predecessor.

paulwalker-arm reviewed (commented):
I don't know enough to know if maintaining the block ordering is what we need/want from a design point of view (it seems kind of fragile to me) but I've added some comments assuming the efficiency gains are worth it.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1005,12 +1005,14 @@ static void replaceVPBBWithIRVPBB(VPBasicBlock *VPBB, BasicBlock *IRBB) {
-    VPBlockUtils::connectBlocks(IRVPBB, Succ);
-    VPBlockUtils::disconnectBlocks(VPBB, Succ);
-  }
+  PredVPBB->replaceSuccessor(VPBB, IRVPBB);
```
Might be good to also go ahead and adjust this code to work for any number of predecessor by iterating over them and replacing the successor?

Appending the predecessors to IRVPBB should be fine, as they should be empty to start with

fhahn reviewed (commented):
Thanks for putting up the patch! 

The order of successors and predecessors is meaningful in VPlan, e.g. to determine the order of operands in a branch instruction for VPBlocks with multiple successors. So it should definitely be preserved here, it wasn't an issue so far as the middle block always had a single predecessor

david-arm committed:
Address review comments

* Add new VPBlockUtils::reassociateBlocks that replaces all
associations to Old with New.
* Removed code to clear successors/predecessors in the deleted
block.

david-arm commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1005,12 +1005,14 @@ static void replaceVPBBWithIRVPBB(VPBasicBlock *VPBB, BasicBlock *IRBB) {
-    VPBlockUtils::disconnectBlocks(VPBB, Succ);
-  }
+  PredVPBB->replaceSuccessor(VPBB, IRVPBB);
+  IRVPBB->setPredecessors({PredVPBB});
```
Good suggestion - I've added a new VPBlockUtils::reassociateBlock function and do the replacements in there!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -556,6 +556,26 @@ class VPBlockBase {
+    assert(I != Predecessors.end());
+    assert(Old->getParent() == New->getParent() &&
+           "replaced predecessor must have the same parent");
+    *I = New;
```
I think this may  leaves `Old` in an invalid state, as we remove it from the list of predecessors here, but don't remove `this` from the `Successors` of `Old`. Can we also remove it from there? Otherwise the callers of the API must either remove the node itself  or completely delete the block

david-arm replied:
Hmm, but if we go down that route then surely we should also care about setting the successor of `New` to `this` as well so that we leave it in a valid state? I guess if this function was private then it wouldn't be part of the interface so it wouldn't matter as much. If we do the extra work here as you suggest, then this patch becomes less about efficiency and more about retaining ordering, which is fine but then I probably ought to change the title!

david-arm replied:
For example, `removeSuccessor` does not leave the `Successor` argument in a valid state either because `this` will still be a predecessor of `Successor`. How about I make the functions private and then we can behave like `removeSuccessor`, unless you see value in these new interfaces being public?

fhahn replied:
Right, let me take a look tomorrow

david-arm replied:
Sure, no problem. I just added a new commit as I realised the the public `reassociateBlocks` interface was leaving `Old` and `New` in an invalid state. Hopefully, this now addresses your concern!

fhahn replied:
Thanks this was what I was about to suggest, as this ensures that `reassociateBlocks` leaves the CFG in a valid state.

david-arm committed:
Make new replaceXYZ functions private

* Made functions private in same way as removeSuccessor and
removePredecessor so that we no longer have to worry about
maintaining a valid state of `Old`.

david-arm committed:
Ensure reassociateBlocks leaves Old and New in a decent state

* Since reassociateBlocks is a public interface we should leave
the Old and New blocks in a usable state.

fhahn reviewed (approved):
LGTM, thanks!

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3862,6 +3882,19 @@ class VPBlockUtils {
+  /// Reassociate all the blocks connected to \p Old so that they now point to
+  /// \p New.
+  static void reassociateBlocks(VPBlockBase *Old, VPBlockBase *New) {
+    for (auto *Pred : to_vector(Old->getPredecessors()))
```
Is this to_vector() needed, given that Old's predecessors remain intact?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3862,6 +3882,19 @@ class VPBlockUtils {
+  static void reassociateBlocks(VPBlockBase *Old, VPBlockBase *New) {
+    for (auto *Pred : to_vector(Old->getPredecessors()))
+      Pred->replaceSuccessor(Old, New);
+    for (auto *Succ : to_vector(Old->getSuccessors()))
```
Is this to_vector() needed, given that Old's successors remain intact?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3862,6 +3882,19 @@ class VPBlockUtils {
   }
 
+  /// Reassociate all the blocks connected to \p Old so that they now point to
+  /// \p New.
```
```suggestion
  /// \p New instead.
```

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -440,6 +440,26 @@ class VPBlockBase {
+  /// trying to replace an old block in the CFG with a new one.
+  void replaceSuccessor(VPBlockBase *Old, VPBlockBase *New) {
+    auto I = find(Successors, Old);
+    assert(I != Successors.end());
```
Error message missing.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1004,13 +1004,9 @@ static void replaceVPBBWithIRVPBB(VPBasicBlock *VPBB, BasicBlock *IRBB) {
     assert(!R.isPhi() && "Tried to move phi recipe to end of block");
```
While we're here, worth adding in documentation above that VPBB, having a single predecessor, is expected to be free of phi recipes. A more informative error message may be "VPBB to be replaced by IRBB must be free of phi recipes".

Note that, in general, replacing a VPBB with an IRBB is a temporary solution, until the full skeleton is built in VPlan. Blocks would clearly be identified as IRBB's or not, from the outset.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1004,13 +1004,9 @@ static void replaceVPBBWithIRVPBB(VPBasicBlock *VPBB, BasicBlock *IRBB) {
     assert(!R.isPhi() && "Tried to move phi recipe to end of block");
     R.moveBefore(*IRVPBB, IRVPBB->end());
   }
-  VPBlockBase *PredVPBB = VPBB->getSinglePredecessor();
```
This assert of having a single predecessor is now lost, yet still documented. The function traverses multiple predecessors, but probably works and tested for single predecessor only, as VPBB is still expected to be free of phi recipes? Better be clear, assert and test what is supported.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -440,6 +440,26 @@ class VPBlockBase {
+  /// trying to replace an old block in the CFG with a new one.
+  void replacePredecessor(VPBlockBase *Old, VPBlockBase *New) {
+    auto I = find(Predecessors, Old);
+    assert(I != Predecessors.end());
```
Error message missing.

ayalz reviewed (commented):
Various post-commit comments.
